### PR TITLE
wasm pkg: add OCI packaged Wasm binary fetcher.

### PR DESCRIPTION
**Pull request by mathetake**
Added the Proxy-Wasm OCI image fetcher in wasm pkg. The OCI image is only specified [here](https://docs.google.com/document/d/1Uf-3RPFRyNl6ExYUeZERlB9Y8kyxZFaB6NbDZyguXwI/edit).

the actual use case of this fetcher will come with the implementation of Wasm Extension API(https://github.com/istio/api/pull/1940).

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

**Comment by bianpengyuan** [comment]
Do you have a rough estimate the binary size will increase by how much after importing this?

File: pkg/wasm/ocifetcher.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		return nil, fmt.Errorf("number of layers must be one but got %d", len(layers))
+	}
+
+	layer := layers[0]
+	mt, err := layer.MediaType()
+	if err != nil {
+		return nil, fmt.Errorf("could not get media type: %v", err)
+	}
+
+	if types.OCIUncompressedLayer != mt {
```
Comment (howardjohn) [inline pkg/wasm/ocifetcher.go:84]: are we only going to support OCIUncompressedLayer long term? are there registries that always use compression, etc?

File: pkg/wasm/ocifetcher.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch layers: %v", err)
+	}
+
+	// The image must be single-layered.
+	if len(layers) != 1 {
+		return nil, fmt.Errorf("number of layers must be one but got %d", len(layers))
+	}
+
+	layer := layers[0]
```
Comment (howardjohn) [inline pkg/wasm/ocifetcher.go:78]: does this limit what we can support at all?

File: pkg/wasm/ocifetcher.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+	// TODO(mathetake): Should we have a basic binary check? (e.g. Wasm header in binary.)
+	// TODO(mathetake): Impl signature verification.
+
+	return ret, nil
+}
+
+// Extracts the Wasm extension file in the archive named "extension.wasm".
+func extractWasmExtensionFile(r *tar.Reader) ([]byte, error) {
+	const wasmExtensionFileName = "extension.wasm"
```
Comment (howardjohn) [inline pkg/wasm/ocifetcher.go:108]: This seems a bit restrictive? Is there a reason we enforce the name?

Comment (mathetake) [reply pkg/wasm/ocifetcher.go:108]: This is following [the spec](https://docs.google.com/document/d/1Uf-3RPFRyNl6ExYUeZERlB9Y8kyxZFaB6NbDZyguXwI/edit) we (@dgn knows best) discussed in Wasm-SIG.
While I agree with that this is a bit restrictive, all the users would use a tool (supposedly developed later) by Proxy-Wasm community to package raw Wasm binaries and they shouldn't care about the actual file name inside resulting OCI images.

File: pkg/wasm/ocifetcher.go (hunk 2 above)
Comment (mathetake) [reply pkg/wasm/ocifetcher.go:78]: this is a specification in https://docs.google.com/document/d/1Uf-3RPFRyNl6ExYUeZERlB9Y8kyxZFaB6NbDZyguXwI/edit

File: pkg/wasm/ocifetcher.go (hunk 3 above)
Comment (mathetake) [reply pkg/wasm/ocifetcher.go:108]: btw, I mistakenly use `extension.wasm` here, not `filter.wasm` in the spec.. will fix.

File: pkg/wasm/ocifetcher.go (hunk 1 above)
Comment (mathetake) [reply pkg/wasm/ocifetcher.go:84]: yeah, that's the point we should discuss.. @dgn could you tell me why you folks who discussed the spec chose `application/vnd.oci.image.layer.v1.tar` rather that its compressed variant `application/vnd.oci.image.layer.v1.tar+gzip`?

**Comment by mathetake** [comment]
> Do you have a rough estimate the binary size will increase by how much after importing this?

let me have a look..

File: pkg/wasm/ocifetcher.go (hunk 1 above)
Comment (dgn) [reply pkg/wasm/ocifetcher.go:84]: The spec was just an ad-hoc doc created during a meeting -- I agree the gzip'ed layer probably makes more sense

Comment (dgn) [reply pkg/wasm/ocifetcher.go:84]: The important part is just that it has to be compliant with all registries, which that should also be

**Comment by mathetake** [comment]
OK, I will wait for the spec to be more concrete before I actually do more work on this to be ready for review.

**Pull request closed without merging**